File: backend/src/utils/scoring-engine.ts
import { Institution, HistoricalRecord, CategoryScores, MetricConfig } from '../../../shared/interfaces';

const CATEGORY_WEIGHTS = {
  capitalization: 0.25,
  asset_quality: 0.30,
  profitability: 0.25,
  liquidity: 0.20
};

function scoreMetric(value: number, config: MetricConfig): number {
  const { excellent, good, fair, poor, reverse = false } = config;

  if (reverse) {
    // Lower values are better
    if (value <= excellent) return 10;
    if (value <= good) return 7 + 3 * (good - value) / (good - excellent);
    if (value <= fair) return 5 + 2 * (fair - value) / (fair - good);
    if (value <= poor) return 2 + 3 * (poor - value) / (poor - fair);
    return 1;
  } else {
    // Higher values are better
    if (value >= excellent) return 10;
    if (value >= good) return 7 + 3 * (value - good) / (excellent - good);
    if (value >= fair) return 5 + 2 * (value - fair) / (good - fair);
    if (value >= poor) return 2 + 3 * (value - poor) / (fair - poor);
    return poor > 0 ? 1 + (value / poor) : 1;
  }
}

export function calculateCapitalizationScore(data: HistoricalRecord, useBrokerLogic: boolean = false): number {
  // Bank logic: Traditional Basel III capital ratios
  const bankConfig: Record<string, MetricConfig> = {
    capital_adequacy_ratio: { weight: 0.35, excellent: 15, good: 12, fair: 10, poor: 8 },
    tier1_ratio: { weight: 0.35, excellent: 12, good: 10, fair: 8, poor: 6 },
    leverage_ratio: { weight: 0.30, excellent: 8, good: 6, fair: 5, poor: 3 }
  };

  // Broker-Dealer logic: Focus on net capital and leverage with stricter thresholds
  const brokerConfig: Record<string, MetricConfig> = {
    capital_adequacy_ratio: { weight: 0.40, excellent: 20, good: 16, fair: 12, poor: 10 },
    tier1_ratio: { weight: 0.30, excellent: 15, good: 12, fair: 9, poor: 6 },
    leverage_ratio: { weight: 0.30, excellent: 10, good: 8, fair: 6, poor: 4 }
  };

  const metricsConfig = useBrokerLogic ? brokerConfig : bankConfig;

  let totalScore = 0;
  for (const [metric, config] of Object.entries(metricsConfig)) {
    const value = data[metric as keyof HistoricalRecord] as number;
    const metricScore = scoreMetric(value, config);
    totalScore += metricScore * config.weight;
  }

  return Math.min(10.0, Math.max(1.0, totalScore));
}

export function calculateAssetQualityScore(data: HistoricalRecord, useBrokerLogic: boolean = false): number {
  // Bank logic: Traditional loan quality metrics
  const bankConfig: Record<string, MetricConfig> = {
    npl_ratio: { weight: 0.40, excellent: 0.5, good: 1.0, fair: 2.0, poor: 3.0, reverse: true },
    loan_loss_provisions: { weight: 0.30, excellent: 0.3, good: 0.5, fair: 1.0, poor: 1.5, reverse: true },
    coverage_ratio: { weight: 0.20, excellent: 120, good: 100, fair: 80, poor: 60 },
    asset_classification: { weight: 0.10, excellent: 1, good: 2, fair: 4, poor: 6, reverse: true }
  };

  // Broker-Dealer logic: Focus on customer protection and counterparty risk
  const brokerConfig: Record<string, MetricConfig> = {
    npl_ratio: { weight: 0.30, excellent: 0.3, good: 0.7, fair: 1.5, poor: 2.5, reverse: true },
    loan_loss_provisions: { weight: 0.25, excellent: 0.2, good: 0.4, fair: 0.8, poor: 1.2, reverse: true },
    coverage_ratio: { weight: 0.25, excellent: 150, good: 120, fair: 90, poor: 70 },
    asset_classification: { weight: 0.20, excellent: 0.5, good: 1.5, fair: 3, poor: 5, reverse: true }
  };

  const metricsConfig = useBrokerLogic ? brokerConfig : bankConfig;

  let totalScore = 0;
  for (const [metric, config] of Object.entries(metricsConfig)) {
    const value = data[metric as keyof HistoricalRecord] as number;
    const metricScore = scoreMetric(value, config);
    totalScore += metricScore * config.weight;
  }

  return Math.min(10.0, Math.max(1.0, totalScore));
}

export function calculateProfitabilityScore(data: HistoricalRecord, useBrokerLogic: boolean = false): number {
  // Bank logic: Net interest margin focused
  const bankConfig: Record<string, MetricConfig> = {
    return_on_assets: { weight: 0.25, excellent: 1.5, good: 1.2, fair: 0.8, poor: 0.4 },
    return_on_equity: { weight: 0.25, excellent: 15, good: 12, fair: 8, poor: 4 },
    net_interest_margin: { weight: 0.25, excellent: 4.0, good: 3.5, fair: 2.5, poor: 1.5 },
    cost_to_income_ratio: { weight: 0.15, excellent: 50, good: 60, fair: 70, poor: 80, reverse: true },
    earnings_per_share: { weight: 0.10, excellent: 10, good: 8, fair: 5, poor: 2 }
  };

  // Broker-Dealer logic: Commission and trading revenue focused with higher ROE expectations
  const brokerConfig: Record<string, MetricConfig> = {
    return_on_assets: { weight: 0.20, excellent: 2.0, good: 1.5, fair: 1.0, poor: 0.5 },
    return_on_equity: { weight: 0.30, excellent: 20, good: 15, fair: 10, poor: 5 },
    net_interest_margin: { weight: 0.15, excellent: 3.0, good: 2.5, fair: 1.8, poor: 1.0 },
    cost_to_income_ratio: { weight: 0.20, excellent: 45, good: 55, fair: 65, poor: 75, reverse: true },
    earnings_per_share: { weight: 0.15, excellent: 12, good: 9, fair: 6, poor: 3 }
  };

  const metricsConfig = useBrokerLogic ? brokerConfig : bankConfig;

  let totalScore = 0;
  for (const [metric, config] of Object.entries(metricsConfig)) {
    const value = data[metric as keyof HistoricalRecord] as number;
    const metricScore = scoreMetric(value, config);
    totalScore += metricScore * config.weight;
  }

  return Math.min(10.0, Math.max(1.0, totalScore));
}

export function calculateLiquidityScore(data: HistoricalRecord, useBrokerLogic: boolean = false): number {
  // Bank logic: Basel III liquidity standards
  const bankConfig: Record<string, MetricConfig> = {
    liquidity_coverage_ratio: { weight: 0.35, excellent: 130, good: 115, fair: 105, poor: 100 },
    net_stable_funding_ratio: { weight: 0.35, excellent: 115, good: 110, fair: 105, poor: 100 },
    loan_to_deposit_ratio: { weight: 0.20, excellent: 70, good: 80, fair: 90, poor: 100, reverse: true },
    cash_ratio: { weight: 0.10, excellent: 15, good: 12, fair: 8, poor: 5 }
  };

  // Broker-Dealer logic: Customer reserve requirements and immediate liquidity focus
  const brokerConfig: Record<string, MetricConfig> = {
    liquidity_coverage_ratio: { weight: 0.30, excellent: 150, good: 130, fair: 115, poor: 105 },
    net_stable_funding_ratio: { weight: 0.30, excellent: 120, good: 115, fair: 108, poor: 102 },
    loan_to_deposit_ratio: { weight: 0.15, excellent: 60, good: 75, fair: 85, poor: 95, reverse: true },
    cash_ratio: { weight: 0.25, excellent: 20, good: 15, fair: 10, poor: 6 }
  };

  const metricsConfig = useBrokerLogic ? brokerConfig : bankConfig;

  let totalScore = 0;
  for (const [metric, config] of Object.entries(metricsConfig)) {
    const value = data[metric as keyof HistoricalRecord] as number;
    const metricScore = scoreMetric(value, config);
    totalScore += metricScore * config.weight;
  }

  return Math.min(10.0, Math.max(1.0, totalScore));
}

export function calculateOverallScore(institution: Institution, useBrokerLogic: boolean = false): CategoryScores {
  const latestData = institution.historical_data[institution.historical_data.length - 1];

  const capitalization = calculateCapitalizationScore(latestData, useBrokerLogic);
  const asset_quality = calculateAssetQualityScore(latestData, useBrokerLogic);
  const profitability = calculateProfitabilityScore(latestData, useBrokerLogic);
  const liquidity = calculateLiquidityScore(latestData, useBrokerLogic);

  const overall = 
    capitalization * CATEGORY_WEIGHTS.capitalization +
    asset_quality * CATEGORY_WEIGHTS.asset_quality +
    profitability * CATEGORY_WEIGHTS.profitability +
    liquidity * CATEGORY_WEIGHTS.liquidity;

  return {
    capitalization,
    asset_quality,
    profitability,
    liquidity,
    overall: Math.min(10.0, Math.max(1.0, overall))
  };
}

export function getRatingLabel(score: number): string {
  if (score >= 8.5) return 'Excellent';
  if (score >= 7.0) return 'Very Good';
  if (score >= 5.5) return 'Good';
  if (score >= 4.0) return 'Fair';
  return 'Poor';
}

File: utils/scoring_engine.py
import numpy as np
import pandas as pd

def calculate_overall_score(bank_data):
    """
    Calculate overall credit score based on weighted metrics across all categories.
    Score range: 1-10 (10 being best)
    
    Args:
        bank_data: Dictionary containing bank's historical data and metadata
    """
    
    # Get latest year data
    latest_data = bank_data['historical_data'].iloc[-1]
    
    # Define scoring weights for each category
    category_weights = {
        'capitalization': 0.25,
        'asset_quality': 0.30,
        'profitability': 0.25,
        'liquidity': 0.20
    }
    
    # Calculate category scores
    cap_score = calculate_capitalization_score(latest_data)
    asset_score = calculate_asset_quality_score(latest_data)
    profit_score = calculate_profitability_score(latest_data)
    liquidity_score = calculate_liquidity_score(latest_data)
    
    # Calculate weighted overall score
    overall_score = (
        cap_score * category_weights['capitalization'] +
        asset_score * category_weights['asset_quality'] +
        profit_score * category_weights['profitability'] +
        liquidity_score * category_weights['liquidity']
    )
    
    return min(10.0, max(1.0, overall_score))

def calculate_capitalization_score(data):
    """Calculate capitalization score based on capital ratios."""
    
    # Define benchmarks and weights for capitalization metrics
    metrics_config = {
        'capital_adequacy_ratio': {'weight': 0.35, 'excellent': 15, 'good': 12, 'fair': 10, 'poor': 8},
        'tier1_ratio': {'weight': 0.35, 'excellent': 12, 'good': 10, 'fair': 8, 'poor': 6},
        'leverage_ratio': {'weight': 0.30, 'excellent': 8, 'good': 6, 'fair': 5, 'poor': 3}
    }
    
    total_score = 0
    for metric, config in metrics_config.items():
        value = data[metric]
        
        # Score individual metric (1-10 scale)
        if value >= config['excellent']:
            metric_score = 10
        elif value >= config['good']:
            metric_score = 7 + 3 * (value - config['good']) / (config['excellent'] - config['good'])
        elif value >= config['fair']:
            metric_score = 5 + 2 * (value - config['fair']) / (config['good'] - config['fair'])
        elif value >= config['poor']:
            metric_score = 2 + 3 * (value - config['poor']) / (config['fair'] - config['poor'])
        else:
            metric_score = 1 + (value / config['poor']) if config['poor'] > 0 else 1
        
        total_score += metric_score * config['weight']
    
    return min(10.0, max(1.0, total_score))

def calculate_asset_quality_score(data):
    """Calculate asset quality score (lower NPL and provisions are better)."""
    
    # Define benchmarks and weights for asset quality metrics
    metrics_config = {
        'npl_ratio': {'weight': 0.40, 'excellent': 0.5, 'good': 1.0, 'fair': 2.0, 'poor': 3.0, 'reverse': True},
        'loan_loss_provisions': {'weight': 0.30, 'excellent': 0.3, 'good': 0.5, 'fair': 1.0, 'poor': 1.5, 'reverse': True},
        'coverage_ratio': {'weight': 0.20, 'excellent': 120, 'good': 100, 'fair': 80, 'poor': 60},
        'asset_classification': {'weight': 0.10, 'excellent': 1, 'good': 2, 'fair': 4, 'poor': 6, 'reverse': True}
    }
    
    total_score = 0
    for metric, config in metrics_config.items():
        value = data[metric]
        
        if config.get('reverse', False):  # Lower values are better
            if value <= config['excellent']:
                metric_score = 10
            elif value <= config['good']:
                metric_score = 7 + 3 * (config['good'] - value) / (config['good'] - config['excellent'])
            elif value <= config['fair']:
                metric_score = 5 + 2 * (config['fair'] - value) / (config['fair'] - config['good'])
            elif value <= config['poor']:
                metric_score = 2 + 3 * (config['poor'] - value) / (config['poor'] - config['fair'])
            else:
                metric_score = 1
        else:  # Higher values are better
            if value >= config['excellent']:
                metric_score = 10
            elif value >= config['good']:
                metric_score = 7 + 3 * (value - config['good']) / (config['excellent'] - config['good'])
            elif value >= config['fair']:
                metric_score = 5 + 2 * (value - config['fair']) / (config['good'] - config['fair'])
            elif value >= config['poor']:
                metric_score = 2 + 3 * (value - config['poor']) / (config['fair'] - config['poor'])
            else:
                metric_score = 1
        
        total_score += metric_score * config['weight']
    
    return min(10.0, max(1.0, total_score))

def calculate_profitability_score(data):
    """Calculate profitability score."""
    
    # Define benchmarks and weights for profitability metrics
    metrics_config = {
        'return_on_assets': {'weight': 0.25, 'excellent': 1.5, 'good': 1.2, 'fair': 0.8, 'poor': 0.4},
        'return_on_equity': {'weight': 0.25, 'excellent': 15, 'good': 12, 'fair': 8, 'poor': 4},
        'net_interest_margin': {'weight': 0.25, 'excellent': 4.0, 'good': 3.5, 'fair': 2.5, 'poor': 1.5},
        'cost_to_income_ratio': {'weight': 0.15, 'excellent': 50, 'good': 60, 'fair': 70, 'poor': 80, 'reverse': True},
        'earnings_per_share': {'weight': 0.10, 'excellent': 10, 'good': 8, 'fair': 5, 'poor': 2}
    }
    
    total_score = 0
    for metric, config in metrics_config.items():
        value = data[metric]
        
        if config.get('reverse', False):  # Lower values are better
            if value <= config['excellent']:
                metric_score = 10
            elif value <= config['good']:
                metric_score = 7 + 3 * (config['good'] - value) / (config['good'] - config['excellent'])
            elif value <= config['fair']:
                metric_score = 5 + 2 * (config['fair'] - value) / (config['fair'] - config['good'])
            elif value <= config['poor']:
                metric_score = 2 + 3 * (config['poor'] - value) / (config['poor'] - config['fair'])
            else:
                metric_score = 1
        else:  # Higher values are better
            if value >= config['excellent']:
                metric_score = 10
            elif value >= config['good']:
                metric_score = 7 + 3 * (value - config['good']) / (config['excellent'] - config['good'])
            elif value >= config['fair']:
                metric_score = 5 + 2 * (value - config['fair']) / (config['good'] - config['fair'])
            elif value >= config['poor']:
                metric_score = 2 + 3 * (value - config['poor']) / (config['fair'] - config['poor'])
            else:
                metric_score = 1
        
        total_score += metric_score * config['weight']
    
    return min(10.0, max(1.0, total_score))

def calculate_liquidity_score(data):
    """Calculate liquidity score."""
    
    # Define benchmarks and weights for liquidity metrics
    metrics_config = {
        'liquidity_coverage_ratio': {'weight': 0.35, 'excellent': 130, 'good': 120, 'fair': 110, 'poor': 100},
        'net_stable_funding_ratio': {'weight': 0.35, 'excellent': 120, 'good': 110, 'fair': 105, 'poor': 100},
        'loan_to_deposit_ratio': {'weight': 0.20, 'excellent': 75, 'good': 80, 'fair': 85, 'poor': 90, 'reverse': True},
        'cash_ratio': {'weight': 0.10, 'excellent': 12, 'good': 10, 'fair': 7, 'poor': 5}
    }
    
    total_score = 0
    for metric, config in metrics_config.items():
        value = data[metric]
        
        if config.get('reverse', False):  # Lower values are better
            if value <= config['excellent']:
                metric_score = 10
            elif value <= config['good']:
                metric_score = 7 + 3 * (config['good'] - value) / (config['good'] - config['excellent'])
            elif value <= config['fair']:
                metric_score = 5 + 2 * (config['fair'] - value) / (config['fair'] - config['good'])
            elif value <= config['poor']:
                metric_score = 2 + 3 * (config['poor'] - value) / (config['poor'] - config['fair'])
            else:
                metric_score = 1
        else:  # Higher values are better
            if value >= config['excellent']:
                metric_score = 10
            elif value >= config['good']:
                metric_score = 7 + 3 * (value - config['good']) / (config['excellent'] - config['good'])
            elif value >= config['fair']:
                metric_score = 5 + 2 * (value - config['fair']) / (config['good'] - config['fair'])
            elif value >= config['poor']:
                metric_score = 2 + 3 * (value - config['poor']) / (config['fair'] - config['poor'])
            else:
                metric_score = 1
        
        total_score += metric_score * config['weight']
    
    return min(10.0, max(1.0, total_score))

def get_score_interpretation(score):
    """Return textual interpretation of score."""
    if score >= 8.5:
        return {"rating": "AAA", "description": "Excellent Credit Quality", "color": "green"}
    elif score >= 7.5:
        return {"rating": "AA", "description": "Very Good Credit Quality", "color": "green"}
    elif score >= 6.5:
        return {"rating": "A", "description": "Good Credit Quality", "color": "blue"}
    elif score >= 5.5:
        return {"rating": "BBB", "description": "Fair Credit Quality", "color": "orange"}
    elif score >= 4.5:
        return {"rating": "BB", "description": "Speculative Credit Quality", "color": "orange"}
    elif score >= 3.5:
        return {"rating": "B", "description": "Highly Speculative", "color": "red"}
    else:
        return {"rating": "CCC", "description": "Poor Credit Quality", "color": "red"}

def calculate_category_scores(bank_data):
    """Calculate individual scores for each category."""
    latest_data = bank_data['historical_data'].iloc[-1]
    
    return {
        'capitalization': calculate_capitalization_score(latest_data),
        'asset_quality': calculate_asset_quality_score(latest_data),
        'profitability': calculate_profitability_score(latest_data),
        'liquidity': calculate_liquidity_score(latest_data)
    }
